lib/src/util/dart_type_utilities.dart

```diff
@@ -12,6 +12,55 @@ import 'package:meta/meta.dart';
 typedef AstNodePredicate = bool Function(AstNode node);
 
 class DartTypeUtilities {
+  static EnumLikeClassDescription asEnumLikeClass(ClassElement classElement) {
+    // See discussion: https://github.com/dart-lang/linter/issues/2083
+    //
+
+    // Must be concrete.
+    if (classElement.isAbstract) {
+      return null;
+    }
+
+    // With only private non-factory constructors.
+    for (var cons in classElement.constructors) {
+      if (!cons.isPrivate || cons.isFactory) {
+        return null;
+      }
+    }
+
+    // And 2 or more static const fields whose type is the enclosing class.
+    var enumConstantNames = <String>[];
+    for (var field in classElement.fields) {
+      // Ensure static const.
+      if (field.isSynthetic || !field.isConst || !field.isStatic) {
+        continue;
+      }
+      // Check for type equality.
+      if (field.type != classElement.thisType) {
+        continue;
+      }
+      enumConstantNames.add(field.name);
+    }
+    if (enumConstantNames.length < 2) {
+      return null;
+    }
+
+    // And no subclasses in the defining library.
+    var type = classElement.thisType;
+    var compilationUnit = classElement.library.definingCompilationUnit;
+    for (var cls in compilationUnit.types) {
+      var classType = cls.thisType;
+      do {
+        classType = classType.superclass;
+        if (classType == type) {
+          return null;
+        }
+      } while (classType != null && !classType.isDartCoreObject);
```

Comment:
It isn't necessary to have both conditions (the supertype of `Object` is `null`).